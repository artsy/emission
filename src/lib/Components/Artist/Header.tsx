import React from "react"
import { Dimensions, NativeModules, StyleSheet, TextStyle, View, ViewStyle } from "react-native"
import { createFragmentContainer, graphql } from "react-relay"
import { Schema, Track, track as _track } from "../../utils/track"
const { ARTemporaryAPIModule } = NativeModules
import { Button } from "@artsy/palette"
import colors from "lib/data/colors"
import Headline from "../Text/Headline"
import SerifText from "../Text/Serif"

import { Header_artist } from "__generated__/Header_artist.graphql"

const isPad = Dimensions.get("window").width > 700

interface Props {
  artist: Header_artist
}

interface State {
  following: boolean
  followersCount: number
}

const track: Track<Props, State> = _track

@track()
class Header extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      following: false,
      followersCount: props.artist.counts.follows as number,
    }
  }

  componentDidMount() {
    NativeModules.ARTemporaryAPIModule.followStatusForArtist(this.props.artist.internalID, (error, following) => {
      if (error) {
        // FIXME: Handle error
        console.error("Artist/Header.tsx", error.message)
      }
      this.setState({ following })
    })
  }

  render() {
    const artist = this.props.artist
    return (
      <View style={{ paddingTop: 20 }}>
        <Headline style={[styles.base, styles.headline]}>{artist.name}</Headline>
        {this.renderByline()}
        {this.renderFollowersCount()}
        {this.renderFollowButton()}
      </View>
    )
  }

  renderFollowButton() {
    if (this.state.following !== null) {
      return (
        <View style={styles.followButton}>
          <Button
            variant={this.state.following ? "secondaryOutline" : "primaryBlack"}
            block
            width={100}
            onPress={this.handleFollowChange.bind(this)}
          >
            {this.state.following ? "Following" : "Follow"}
          </Button>
        </View>
      )
    }
  }

  renderFollowersCount() {
    const count = this.state.followersCount
    const followerString = count + (count === 1 ? " Follower" : " Followers")
    return <SerifText style={[styles.base, styles.followCount]}>{followerString}</SerifText>
  }

  renderByline() {
    const artist = this.props.artist
    const bylineRequired = artist.nationality || artist.birthday
    if (bylineRequired) {
      return (
        <View>
          <SerifText style={styles.base}>{this.descriptiveString()}</SerifText>
        </View>
      )
    } else {
      return null
    }
  }

  descriptiveString() {
    const artist = this.props.artist
    const descriptiveString = (artist.nationality || "") + this.birthdayString()
    return descriptiveString
  }

  birthdayString() {
    const birthday = this.props.artist.birthday
    if (!birthday) {
      return ""
    }

    const leadingSubstring = this.props.artist.nationality ? ", b." : ""

    if (birthday.includes("born")) {
      return birthday.replace("born", leadingSubstring)
    } else if (birthday.includes("Est.") || birthday.includes("Founded")) {
      return " " + birthday
    }

    return leadingSubstring + " " + birthday
  }

  @track((props, state) => ({
    action_name: state.following ? Schema.ActionNames.ArtistUnfollow : Schema.ActionNames.ArtistFollow,
    action_type: Schema.ActionTypes.Tap,
    owner_id: props.artist.internalID,
    owner_slug: props.artist.gravityID,
    owner_type: Schema.OwnerEntityTypes.Artist,
  }))
  handleFollowChange() {
    const newFollowersCount = this.state.following ? this.state.followersCount - 1 : this.state.followersCount + 1
    ARTemporaryAPIModule.setFollowArtistStatus(
      !this.state.following,
      this.props.artist.internalID,
      (error, following) => {
        if (error) {
          console.warn(error)
          this.failedFollowChange()
        } else {
          this.successfulFollowChange()
        }
        this.setState({ following, followersCount: newFollowersCount })
      }
    )
    this.setState({ following: !this.state.following, followersCount: newFollowersCount })
  }

  @track((props, state) => ({
    action_name: state.following ? Schema.ActionNames.ArtistFollow : Schema.ActionNames.ArtistUnfollow,
    action_type: Schema.ActionTypes.Success,
    owner_id: props.artist.internalID,
    owner_slug: props.artist.gravityID,
    owner_type: Schema.OwnerEntityTypes.Artist,
  }))
  successfulFollowChange() {
    // callback for analytics purposes
  }

  @track((props, state) => ({
    action_name: state.following ? Schema.ActionNames.ArtistFollow : Schema.ActionNames.ArtistUnfollow,
    action_type: Schema.ActionTypes.Fail,
    owner_id: props.artist.internalID,
    owner_slug: props.artist.gravityID,
    owner_type: Schema.OwnerEntityTypes.Artist,
  }))
  failedFollowChange() {
    // callback for analytics purposes
  }
}

interface Styles {
  base: TextStyle
  headline: TextStyle
  followCount: TextStyle
  followButton: ViewStyle
}

const styles = StyleSheet.create<Styles>({
  base: {
    textAlign: "center",
  },
  headline: {
    fontSize: 14,
  },
  followCount: {
    color: colors["gray-semibold"],
    marginBottom: 30,
  },
  followButton: {
    height: 40,
    width: isPad ? 330 : null,
    alignSelf: isPad ? "center" : null,
    marginLeft: 0,
    marginRight: 0,
  },
})

export default createFragmentContainer(Header, {
  artist: graphql`
    fragment Header_artist on Artist {
      internalID
      slug
      name
      nationality
      birthday
      counts {
        follows
      }
    }
  `,
})
