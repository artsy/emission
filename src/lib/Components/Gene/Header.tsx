import React from "react"
import { Dimensions, NativeModules, StyleSheet, TextStyle, View, ViewStyle } from "react-native"
import { createFragmentContainer, graphql } from "react-relay"
const { ARTemporaryAPIModule } = NativeModules
import { Button } from "@artsy/palette"
import { Schema, Track, track as _track } from "../../utils/track"
import Headline from "../Text/Headline"

import { Header_gene } from "__generated__/Header_gene.graphql"

interface Props {
  gene: Header_gene
  shortForm: boolean
}

interface State {
  following: boolean | null
}

const track: Track<Props, State> = _track

@track()
class Header extends React.Component<Props, State> {
  state = { following: null }

  componentDidMount() {
    NativeModules.ARTemporaryAPIModule.followStatusForGene(this.props.gene.internalID, (error, following) => {
      if (error) {
        // FIXME: Handle error
        console.error("Gene/Header.tsx", error.message)
      }
      this.setState({ following })
    })
  }

  render() {
    const gene = this.props.gene
    return (
      <View>
        <View style={styles.header}>
          <Headline style={styles.headline} numberOfLines={2}>
            {gene.name}
          </Headline>
        </View>
        {this.renderFollowButton()}
      </View>
    )
  }

  @track((props, state) => ({
    action_name: state.following ? Schema.ActionNames.GeneUnfollow : Schema.ActionNames.GeneFollow,
    action_type: Schema.ActionTypes.Tap,
    owner_id: props.gene.internalID,
    owner_slug: props.gene.gravityID,
    owner_type: Schema.OwnerEntityTypes.Gene,
  }))
  handleFollowChange() {
    ARTemporaryAPIModule.setFollowGeneStatus(!this.state.following, this.props.gene.internalID, (error, following) => {
      if (error) {
        console.warn(error)
        this.failedFollowChange()
      } else {
        this.successfulFollowChange()
      }
      this.setState({ following })
    })
    this.setState({ following: !this.state.following })
  }

  @track((props, state) => ({
    action_name: state.following ? Schema.ActionNames.GeneFollow : Schema.ActionNames.GeneUnfollow,
    action_type: Schema.ActionTypes.Success,
    owner_id: props.gene.internalID,
    owner_slug: props.gene.gravityID,
    owner_type: Schema.OwnerEntityTypes.Gene,
  }))
  successfulFollowChange() {
    // callback for analytics purposes
  }

  @track((props, state) => ({
    action_name: state.following ? Schema.ActionNames.GeneFollow : Schema.ActionNames.GeneUnfollow,
    action_type: Schema.ActionTypes.Fail,
    owner_id: props.gene.internalID,
    owner_slug: props.gene.gravityID,
    owner_type: Schema.OwnerEntityTypes.Gene,
  }))
  failedFollowChange() {
    // callback for analytics purposes
  }

  renderFollowButton() {
    if (this.props.shortForm) {
      return null
    }
    return (
      <View style={styles.followButton}>
        <Button
          variant={this.state.following ? "secondaryOutline" : "primaryBlack"}
          block
          width={100}
          onPress={this.handleFollowChange.bind(this)}
        >
          {this.state.following ? "Following" : "Follow"}
        </Button>
      </View>
    )
  }
}

const isPad = Dimensions.get("window").width > 700

interface Styles {
  header: TextStyle
  headline: TextStyle
  followButton: ViewStyle
}

const styles = StyleSheet.create<Styles>({
  header: {
    marginTop: 15,
    height: 36,
    justifyContent: "center",
    alignItems: "center",
    paddingLeft: 40,
    paddingRight: 40,
  },
  headline: {
    textAlign: "center",
    fontSize: isPad ? 20 : 14,
  },
  followButton: {
    height: 40,
    marginTop: 30,
  },
})

export default createFragmentContainer(Header, {
  gene: graphql`
    fragment Header_gene on Gene {
      internalID
      slug
      name
    }
  `,
})
