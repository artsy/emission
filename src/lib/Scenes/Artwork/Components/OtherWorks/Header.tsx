import { Flex, Serif } from "@artsy/palette"
import React from "react"

interface HeaderProps {
  children?: JSX.Element
  title: string
}

export const Header: React.SFC<HeaderProps> = props => {
  const { children, title } = props

  return (
    <Flex flexDirection="column" alignItems="center">
      <Serif size="5t" mb={2} textAlign="center">
        {title}
      </Serif>
      {children}
    </Flex>
  )
}
