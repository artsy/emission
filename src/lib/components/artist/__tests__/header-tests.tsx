import { NativeModules } from 'react-native'

import * as React from 'react'
import * as renderer from 'react-test-renderer'

import Header from '../header'

// jest.mock('../../spinner.tsx', () => 'ARSpinner')

beforeAll(() => {
  NativeModules.ARTemporaryAPIModule = { followStatusForArtist: jest.fn() }
})

it('renders properly', () => {
  const artist = {
      name: 'Marcel Duchamp',
      nationality: 'French',
      birthday: '11/17/1992',
      counts: {
        follows: 22,
      }
  }
  const header = renderer.create(<Header artist={artist} />).toJSON()
  expect(header).toMatchSnapshot()
})
