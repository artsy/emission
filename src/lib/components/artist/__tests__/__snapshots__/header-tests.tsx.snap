exports[`test renders properly 1`] = `
<Header
  artist={
    Object {
      "birthday": "11/17/1992",
      "counts": Object {
        "follows": 22,
      },
      "name": "Marcel Duchamp",
      "nationality": "French",
    }
  } />
`;
